src/devices/src/legacy/serial.rs

```diff
@@ -0,0 +1,214 @@
+        data[0] = self.0.read(offset);
+    }
+
+    fn bus_write(&mut self, offset: u8, data: &[u8]) {
+        if data.len() != 1 {
+            debug!(
+                "Serial console invalid data length on write: {}",
+                data.len()
+            );
+            return;
+        }
+
+        // This is safe because we checked that `data` has length 1.
+        let res = self.0.write(offset, data[0]);
+        if res.is_err() {
+            debug!("Error writing to serial console: {:#?}", res.unwrap_err());
+        }
+    }
+}
+
+#[cfg(target_arch = "x86_64")]
+impl<T: Trigger<E = io::Error>, W: Write> MutDevicePio for SerialWrapper<T, NoEvents, W> {
+    fn pio_read(&mut self, _base: PioAddress, offset: PioAddressValue, data: &mut [u8]) {
+        // TODO: this function can't return an Err, so we'll mark error conditions
+        // (data being more than 1 byte, offset overflowing an u8) with logs & metrics.
+
+        match offset.try_into() {
+            Ok(offset) => self.bus_read(offset, data),
+            Err(_) => debug!("Invalid serial console read offset."),
+        }
+    }
+
+    fn pio_write(&mut self, _base: PioAddress, offset: PioAddressValue, data: &[u8]) {
+        // TODO: this function can't return an Err, so we'll mark error conditions
+        // (data being more than 1 byte, offset overflowing an u8) with logs & metrics.
+
+        match offset.try_into() {
+            Ok(offset) => self.bus_write(offset, data),
+            Err(_) => debug!("Invalid serial console write offset."),
+        }
+    }
+}
+
+#[cfg(target_arch = "aarch64")]
+impl<T: Trigger<E = io::Error>, W: Write> MutDeviceMmio for SerialWrapper<T, NoEvents, W> {
+    fn mmio_read(&mut self, _base: MmioAddress, offset: u64, data: &mut [u8]) {
+        // TODO: this function can't return an Err, so we'll mark error conditions
+        // (data being more than 1 byte, offset overflowing an u8) with logs & metrics.
+
+        match offset.try_into() {
+            Ok(offset) => self.bus_read(offset, data),
+            Err(_) => debug!("Invalid serial console read offset."),
+        }
+    }
+
+    fn mmio_write(&mut self, _base: MmioAddress, offset: u64, data: &[u8]) {
+        // TODO: this function can't return an Err, so we'll mark error conditions
+        // (data being more than 1 byte, offset overflowing an u8) with logs & metrics.
+
+        match offset.try_into() {
+            Ok(offset) => self.bus_write(offset, data),
+            Err(_) => debug!("Invalid serial console write offset."),
+        }
+    }
+}
+/// Errors encountered during device operation.
+#[derive(Debug)]
+pub enum Error {
+    /// Failed to create an event manager for device events.
+    EventManager(event_manager::Error),
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::legacy::EventFdTrigger;
+
+    use std::io::sink;
+
+    fn test_invalid_data(invalid_data: &mut [u8]) {
```

Comment:
nit: rename to `check_` or some other name. We reserve `test_` for tests.

Reply:
Renamed this.